### PR TITLE
Add support for FIRST_VALUE, LAST_VALUE Aggregate Functions

### PR DESCRIPTION
**Pull request by mustafasrepo**
# Which issue does this PR close?

Closes [#6444](https://github.com/apache/arrow-datafusion/issues/6444).

# Rationale for this change

# What changes are included in this PR?
This PR adds support for `FIRST_VALUE`, `LAST_VALUE` ordering sensitive aggregation functions. With the changes in this PR we now can run query below
```sql
SELECT a, FIRST_VALUE(c ORDER BY b DESC) as first_c
FROM table
GROUP BY a
```
This query will return the first c value, according to the reverse b order, for every a group.

# Are these changes tested?
Yes, new tests are added to the `groupby.slt` file for new aggregate functions.

# Are there any user-facing changes?

### CODE REVIEW

**Comment by ozankabak** [comment]
Note that this change makes `FIRST_VALUE` and `LAST_VALUE` functions, which are already available as window functions, also available in an aggregation context so that one can choose first/last values within a group according to a given ordering. These functions are sometimes called `FIRST` and `LAST` in other query engines/DBs, but we chose to use the same with their window function counterparts (as in the case of `SUM`, `COUNT` etc.)

Having said that, there are few minor issues here that we will fix shortly. I think the example in the PR body should read something like:
```sql
SELECT a, FIRST_VALUE(c ORDER BY b DESC) as first_c
FROM table
GROUP BY a
```
This will return the first `c` value, according to the reverse `b` order, for every `a` group.

Also some tests are failing. We will fix the tests, make the PR body clearer and possibly add another test that looks like this.

**Comment by alamb** [comment]
Marking as draft while the PR is worked on

File: datafusion/core/tests/dataframe_functions.rs [hunk 1]
```diff
@@ -155,11 +155,11 @@ async fn test_fn_approx_median() -> Result<()> {
     let expr = approx_median(col("b"));
 
     let expected = vec![
-        "+----------------------+",
-        "| APPROXMEDIAN(test.b) |",
-        "+----------------------+",
-        "| 10                   |",
-        "+----------------------+",
+        "+-----------------------+",
```
Comment (mustafasrepo) [inline datafusion/core/tests/dataframe_functions.rs:158]: As part of this PR, I have added `convert_camel_to_upper_snake` util for display of Aggregate function. Their display name is automatically, calculate from struct name. Hence display name of some of the existing aggregators is changed.

**Comment by mustafasrepo** [comment]
> Marking as draft while the PR is worked on

I have resolved the CI issue, also I have updated Pr body to reflect intent better. This PR is ready for further review.

File: datafusion/core/tests/sqllogictests/test_files/window.slt [hunk 2]
```diff
@@ -3017,6 +3017,13 @@ SELECT a, b, c,
 0 0 3 11 96 11 2 10 36 10 36 11 5 11 9
 0 0 4 9 72 9 NULL 14 45 14 45 9 4 9 9
 
+#fn aggregate order by with window frame
```
Comment (alamb) [inline datafusion/core/tests/sqllogictests/test_files/window.slt:3020]: 👍

File: datafusion/core/tests/sqllogictests/test_files/window.slt [hunk 3]
```diff
@@ -3017,6 +3017,13 @@ SELECT a, b, c,
...
+# In window expressions, aggregate functions should not have an ordering requirement, such requirements
+# should be defined in the window frame. Therefore, the query below should generate an error. Note that
+# PostgreSQL also behaves this way.
+statement error DataFusion error: Error during planning: Aggregate ORDER BY is not implemented for window functions
+SELECT SUM(inc_col ORDER BY a DESC) OVER() as last
+       FROM annotated_data_infinite2
```
Comment (alamb) [inline datafusion/core/tests/sqllogictests/test_files/window.slt:3026]: I recommend another negative test to make sure the window function and over clause can not be used in the same query. Something like

```sql
EXPLAIN SELECT a, b, LAST_VALUE(c ORDER BY a DESC) OVER (order by a) as last_c
```

File: datafusion/core/tests/sqllogictests/test_files/groupby.slt [hunk 4]
```diff
@@ -2007,6 +2008,58 @@ SELECT a, d,
...
+logical_plan
+Projection: annotated_data_infinite2.a, annotated_data_infinite2.b, LAST_VALUE(annotated_data_infinite2.c) ORDER BY [annotated_data_infinite2.a DESC NULLS FIRST] AS last_c
+--Aggregate: groupBy=[[annotated_data_infinite2.a, annotated_data_infinite2.b]], aggr=[[LAST_VALUE(annotated_data_infinite2.c) ORDER BY [annotated_data_infinite2.a DESC NULLS FIRST]]]
+----TableScan: annotated_data_infinite2 projection=[a, b, c]
+physical_plan
+ProjectionExec: expr=[a@0 as a, b@1 as b, LAST_VALUE(annotated_data_infinite2.c) ORDER BY [annotated_data_infinite2.a DESC NULLS FIRST]@2 as last_c]
+--AggregateExec: mode=Single, gby=[a@0 as a, b@1 as b], aggr=[LAST_VALUE(annotated_data_infinite2.c)], ordering_mode=FullyOrdered
+----CsvExec: file_groups={1 group: [[WORKSPACE_ROOT/datafusion/core/tests/data/window_2.csv]]}, projection=[a, b, c], infinite_source=true, output_ordering=[a@0 ASC NULLS LAST, b@1 ASC NULLS LAST, c@2 ASC NULLS LAST], has_header=true
+
+query III
```
Comment (alamb) [inline datafusion/core/tests/sqllogictests/test_files/groupby.slt:2053]: I wonder if it is worth adding some sort of test of the behavior when there is no order? Like `SELECT a, b, LAST_VALUE(c) as last_c` -- mostly to ensure it doesn't cause an error.

**Review by alamb — APPROVED** [review]
Thank you @mustafasrepo  -- I think the feature looks great and really nicely coded 👌 . 

I think a different approach for display name that is not so tightly bound to the Rust struct name would be better, but that is a stylistic opinion and something we can fix in a follow on PR. I left some suggestions to that effect

File: datafusion/expr/src/aggregate_function.rs [hunk 5]
```diff
@@ -79,7 +84,9 @@ pub enum AggregateFunction {
 impl fmt::Display for AggregateFunction {
     fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
         // uppercase of the debug.
-        write!(f, "{}", format!("{self:?}").to_uppercase())
+        // Convert Camel form to uppercase snake
+        // such as FirstValue => FIRST_VALUE
+        write!(f, "{}", convert_camel_to_upper_snake(format!("{self:?}")))
     }
```
[the remark this answers is not in the document]
Comment (mustafasrepo) [reply datafusion/expr/src/aggregate_function.rs:90]: I think, your suggestion is better. By this way, we can avoid to overfit, of enum name convention. Also, having explicit match, helps with debugging also. I will add  your suggestion to this Pr also.

File: datafusion/core/tests/sqllogictests/test_files/groupby.slt (hunk 4 above)
Comment (mustafasrepo) [reply datafusion/core/tests/sqllogictests/test_files/groupby.slt:2053]: I added a test for this use case.

File: datafusion/core/tests/sqllogictests/test_files/window.slt (hunk 3 above)
Comment (mustafasrepo) [reply datafusion/core/tests/sqllogictests/test_files/window.slt:3032]: I added a test for this use case.

File: datafusion/expr/src/aggregate_function.rs [hunk 6]
```diff
@@ -81,12 +80,43 @@ pub enum AggregateFunction {
     BoolOr,
 }
 
+impl AggregateFunction {
```
Comment (alamb) [inline datafusion/expr/src/aggregate_function.rs:83]: 👍

**Comment by alamb** [comment]
Thanks again @mustafasrepo !

**Pull request merged**